File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        

        hn = len(h)

        lw = [0]*hn
        rw = [0]*hn 
        for l in range(1,hn):
            lw[l] = max( lw[l-1] , h[l-1] )

        for r in range(hn-2,-1,-1):
            rw[r] = max( rw[r+1] , h[r+1] ) 

        rain = 0 
        for i in range(hn): 
            wlev = min( lw[i] , rw[i] ) 
            if wlev > h[i]:
                rain += wlev -h[i] 

        return rain 
    
    

File: 42-trapping-rain-water/42-trapping-rain-water.go
func trap(h []int) int {
    
    max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}

	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

    
	hn := len(h)
	lw := make([]int, hn)
	rw := make([]int ,hn) 
	for l := 1 ; l < hn ; l++ { 
        lw[l] = max( lw[l-1] , h[l-1])
		}
	
	for r:=hn-2 ; r >=0; r-- { 
		rw[r] = max( rw[r+1] , h[r+1] )
	}

	ans := 0 
	for i := range h {
        wlev := min( lw[i] , rw[i] )
		if wlev > h[i] { 
			ans += wlev - h[i] }
	}
	return ans 


    
    
    
    
}